src/librbd/ImageWatcher.cc

```diff
@@ -121,8 +131,19 @@ int ImageWatcher::unregister_watch() {
 
   cancel_async_requests(-ESHUTDOWN);
 
-  RWLock::WLocker l(m_watch_lock);
-  return m_image_ctx.md_ctx.unwatch2(m_handle);
+  int r = 0;
+  {
+    RWLock::WLocker l(m_watch_lock);
+    assert(m_watch_state != WATCH_STATE_UNREGISTERED);
+    if (m_watch_state == WATCH_STATE_REGISTERED) {
```

Comment:
How about with my watch patch? We're unwatching watches with errors anyway
when they're not immediately unregistered
On Feb 3, 2015 6:03 PM, "Jason Dillaman" <notifications@github.com> wrote:

> In src/librbd/ImageWatcher.cc
> <https://github.com/ceph/ceph/pull/3592#discussion_r24019610>:
>
> > @@ -121,8 +131,19 @@ int ImageWatcher::unregister_watch() {
> >
> >    cancel_async_requests(-ESHUTDOWN);
> >
> > -  RWLock::WLocker l(m_watch_lock);
> > -  return m_image_ctx.md_ctx.unwatch2(m_handle);
> > +  int r = 0;
> > +  {
> > +    RWLock::WLocker l(m_watch_lock);
> > +    assert(m_watch_state != WATCH_STATE_UNREGISTERED);
> > +    if (m_watch_state == WATCH_STATE_REGISTERED) {
>
> OK, I'll look into it more. I am pretty sure I was originally getting a
> seg fault within librados when I attempted to unwatch a handle that had the
> error handler invoked.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/3592/files#r24019610>.
>